### ISSUE TITLE
Mojom bindings generators should handle "keywords" for their target languages

### ISSUE THREAD

**Issue by viettrungluu-cr** [domokit/mojo#591]
Generators should generate valid code for all valid .mojom files.

In particular, names in .mojom files may coincide with (or get mapped to things that coincide with) keywords and such in various languages. Generators need to map these valid names.

We should have tests for these. Note that we need to test various forms of these (e.g., "IF", "if", "If") in various situations (message names, field/argument names, enum value names, struct/enum/interface/union type names, etc.), since there are different mappings.

(In C++, presumably we also need to avoid conflicting with at least standard macros. Unfortunately, this also may include some lowercase things, like "errno", "assert", "stdin", etc.)

/cc @vardhan-cr @AlexandreZani

**Comment by viettrungluu-cr** [comment]
(For C++, I also forgot to mention context-dependent keywords.)

**Comment by j9brown** [comment]
How do you propose to resolve identifier collisions?

**Comment by jamesr** [comment]
Each language will need to define mangling rules. We can try to establish some general patterns but the details will vary depending on the target language's grammar and conventions.

**Comment by kulakowski** [comment]
Another possible point in the design space is to be able to override the generated names in an annotation. Not sure that's very nice, though, especially if it's the only mechanism. In particular, it's brittle to the author only using X and Y language and not caring about Z.

**Comment by jamesr** [comment]
We can't really expect authors to know the naming rules for all language X generator combinations that might be used with that mojom interface for all time.

**Comment by j9brown** [comment]
I like the suggestion of an annotation as an alternative to automatic name
mangling, to allow an author to selectively override the mangled names for
particular languages.

Note that name mangling also introduces the potential for provoking naming
collisions, depending on how it is implemented.

**Comment by jamesr** [comment]
An annotation scheme can supplement, but not replace, a mangling scheme.  The mangling scheme has to also take into account name collisions.

**Comment by krisgiesing** [comment]
Right, it seems reasonable to have the mangler for general collision avoidance, and then annotations as an escape hatch if a developer runs into local issues.
